backend/amd64/simd_selection.ml

```diff
@@ -463,3 +463,347 @@ let () =
   Location.register_error_of_exn (function
     | Error err -> Some (Location.error_of_printer_file report_error err)
     | _ -> None)
+
+type width =
+  | W8
+  | W16
+  | W32
+  | W64
+
+type register =
+  | New of int
+  | Argument of int
+  | Result of int
+  | Original of int
+
+type vectorized_instruction =
+  { operation : Cfg.operation;
+    arguments : register array;
+    results : register array
+  }
+
+let vector_width = 128
+
+let vectorize_operation width (cfg_ops : Cfg.operation list) :
+    vectorized_instruction list option =
+  (* Assumes cfg_ops are isomorphic and can be vectorized *)
+  let length = List.length cfg_ops in
+  assert (length * width = vector_width);
+  let width_type =
+    match width with
+    | 64 -> W64
+    | 32 -> W32
+    | 16 -> W16
+    | 8 -> W8
+    | _ -> assert false
+  in
+  let make_default arg_count res_count operation =
+    Some
+      [ { operation;
+          arguments = Array.init arg_count (fun i -> Argument i);
+          results = Array.init res_count (fun i -> Result i)
+        } ]
+  in
+  let create_const_vec consts =
+    let rec get_list list n =
+      if n = 0
+      then [], list
+      else
+        match list with
+        | [] -> assert false
+        | hd :: tl ->
+          let list1, list2 = get_list tl (n - 1) in
+          hd :: list1, list2
+    in
+    let highs, lows = get_list consts (length / 2) in
+    let pack_int64 nums =
+      let mask = Int64.shift_right_logical Int64.minus_one (64 - width) in
+      List.fold_left
+        (fun target num ->
+          Int64.logor (Int64.shift_left target width) (Int64.logand num mask))
+        0L nums
+    in
+    Cfg.Const_vec128 { high = pack_int64 highs; low = pack_int64 lows }
+    |> make_default 0 1
+  in
+  let add_op =
+    let sse_op =
+      match width_type with
+      | W64 -> Add_i64
+      | W32 -> Add_i32
+      | W16 -> Add_i16
+      | W8 -> Add_i8
+    in
+    Some (Cfg.Specific (Isimd (SSE2 sse_op)))
+  in
+  let mul_op =
+    match width_type with
+    | W64 -> None
+    | W32 -> Some (Cfg.Specific (Isimd (SSE41 Mullo_i32)))
+    | W16 -> Some (Cfg.Specific (Isimd (SSE2 Mullo_i16)))
+    | W8 -> None
+  in
+  let vectorize_intop (intop : Mach.integer_operation) =
+    match intop with
+    | Iadd -> Option.bind add_op (make_default 2 1)
+    | Isub ->
+      let sse_op =
+        match width_type with
+        | W64 -> Sub_i64
+        | W32 -> Sub_i32
+        | W16 -> Sub_i16
+        | W8 -> Sub_i8
+      in
+      Cfg.Specific (Isimd (SSE2 sse_op)) |> make_default 2 1
```

Comment:
Why is this case different from
the ones for `Iadd` / `Imul`?

Reply:
(asking because if we can make
the cases look similar, it will make
review an maintenance easier.)

Reply:
oh, add and mul are reused in lea but not the rest of the functions, should I still pull out all the other instructions like add and mul?